### PR TITLE
Fix support for dataclasses with ClassVar/InitVar in `apply_to_collection`

### PR DESCRIPTION
**Pull request by a-gardner1**
## What does this PR do?

Adds/fixes support for ClassVar and InitVar fields in dataclasses to apply_to_collection

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/9208

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/utilities/test_apply_func.py [hunk 1]
```diff
@@ -115,24 +150,54 @@ class ModelExample:
...
+        assert dataclasses.is_dataclass(actual) and not isinstance(
+            actual, type
+        ), "Reduction of a dataclass should result in a dataclass"
+        assert dataclasses.is_dataclass(actual.feature) and not isinstance(
+            actual.feature, type
+        ), "Reduction of a nested dataclass should result in a nested dataclass"
+        assert (
+            actual.example_ids == expected.example_ids
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert torch.allclose(
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:164]: these should equal, not just "close" :)

File: tests/utilities/test_apply_func.py [hunk 2]
```diff
@@ -115,24 +150,54 @@ class ModelExample:
...
+    assert dataclasses.is_dataclass(reduced["l"]) and not isinstance(
+        reduced["l"], type
+    ), "Reduction of a dataclass with class and init-only vars should result in a dataclass"
+    assert torch.equal(
+        WithClassAndInitVar.class_var, torch.tensor(0)
+    ), "Reduction of a dataclass with class and init-only vars should not change the class var"
+    try:
+        _assert_nested_dataclass_reduction(reduced["l"].dummy, expected_result["l"].dummy)
+    except AssertionError:
+        raise AssertionError("Reduction of a dataclass with class and init-only vars did not yield the desired result")
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:200]:
```suggestion
    _assert_nested_dataclass_reduction(reduced["l"].dummy, expected_result["l"].dummy)
```
what's the point of that? all the messages there get discarded for this?

File: tests/utilities/test_apply_func.py [hunk 3]
```diff
@@ -115,24 +150,54 @@ class ModelExample:
...
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert torch.allclose(
+            actual.label, expected.label
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert torch.allclose(
+            actual.feature.input_ids, expected.feature.input_ids
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert np.allclose(
+            actual.feature.segment_ids, expected.feature.segment_ids
+        ), "Reduction of a nested dataclass did not yield the desired result"
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:172]: what is the point of these messages? isn't it obvious that something did not yield the desired result when this assertion fails? the function is called `_assert_nested_dataclass_reduction` which should already give a clear idea that we are dealing with nested dataclasses here, and the message is just repeated. it does not add to the readability of the test.

**Review by awaelchli — APPROVED** [review]

File: tests/utilities/test_apply_func.py (hunk 1 above)
Comment (a-gardner1) [reply tests/utilities/test_apply_func.py]: Since these tensors were float like all of the rest, I interpreted this comment to mean that all applications of `allclose` should be replaced with `equal` and have done so throughout the test.

File: tests/utilities/test_apply_func.py (hunk 2 above)
Comment (a-gardner1) [reply tests/utilities/test_apply_func.py]: The main intention was to highlight that the test failed for a dataclass with class and init vars, presuming that if it had gotten this far it must have succeeded for those dataclasses without.
I introduced the `_assert_nested_dataclass_reduction` to avoid duplicating the preexisting tests.

I agree it was a bit awkward and have refactored the dataclass testing mechanisms to minimize duplication and encapsulate dataclass comparison in `__eq__` methods.

File: tests/utilities/test_apply_func.py (hunk 3 above)
Comment (a-gardner1) [reply tests/utilities/test_apply_func.py]: Agreed. These assertions preexisted this PR; I just wrapped them in `_assert_nested_dataclass_reduction`. The latest version cuts down on the message duplication.

File: pytorch_lightning/utilities/apply_func.py [hunk 4]
```diff
@@ -117,10 +117,10 @@ def apply_to_collection(
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
     if _is_dataclass_instance(data):
-        out_dict = {}
-        for field in data.__dataclass_fields__:
+        result = copy(data)
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:120]: Could the copy be avoided?

File: tests/utilities/test_apply_func.py [hunk 5]
```diff
@@ -31,12 +32,80 @@ class Feature:
         input_ids: torch.Tensor
         segment_ids: np.ndarray
 
+        def __eq__(self, o: object) -> bool:
```
Comment (carmocca) [inline tests/utilities/test_apply_func.py:35]: noob q: why did you need to implement all the `__eq__` methods?

Comment (a-gardner1) [reply tests/utilities/test_apply_func.py:35]: They are not strictly necessary, but they avoid repeatedly writing the same or slightly varied checks for each occurrence of one of the dataclasses in the test. A single `==` works for each assertion.

I'll also note that dataclasses are automatically provided a default `__eq__`, but the defaults won't work here since `torch.Tensor` overrides `__eq__` to return tensors instead of bool.

File: pytorch_lightning/utilities/apply_func.py (hunk 4 above)
Comment (a-gardner1) [reply pytorch_lightning/utilities/apply_func.py:120]: Probably, although it might reduce the readability of the code by using lower-level operations. I think `elem_type.__new__()` could be used to instantiate an instance of the same dataclass, but I'd have to test that. 

This is at least a shallow copy, so it is pretty lightweight. Let me know what your preference is.

File: pytorch_lightning/utilities/apply_func.py [hunk 6]
```diff
@@ -129,8 +129,9 @@ def apply_to_collection(
                 **kwargs,
             )
             if include_none or v is not None:
-                out_dict[field] = v
-        return elem_type(**out_dict)
+                setattr(result, field.name, v)
+            # else retain old field value
```
Comment (a-gardner1) [inline pytorch_lightning/utilities/apply_func.py:133]: What should the expected behavior be here?

In other branches of `apply_to_collection`, `include_none=False` leads to the `None`-mapped item being dropped from the collection. I think it would be bad to drop a field from a dataclass, but setting it to `None` might be acceptable (letting it be user error if this violates some invariant for their dataclass).

Alternatively, the original value is retained as it is currently (as though the function was not applied to it).
At the moment, the function will retain the old field value such that `id(getattr(data, field.name)) == id(getattr(result, field.name))`, which seems problematic to me and could lead to unexpected mutability.
I think the proper fix would be to `deepcopy` the field value such that `id(getattr(data, field.name)) != id(getattr(result, field.name))` but `getattr(data, field.name) == getattr(result, field.name)`.

Let me know your preference.

File: tests/utilities/test_apply_func.py (hunk 1 above)
Comment (awaelchli) [reply tests/utilities/test_apply_func.py]: I think your additions to the test should assert equality. There is no reason to assume the value is changed by any operations in `apply_to_collection`.

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -368,6 +368,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed an issue where class or init-only variables of dataclasses were passed to the dataclass constructor in `utilities.apply_to_collection` ([#9208](https://github.com/PyTorchLightning/pytorch-lightning/issues/9208))
```
Comment (awaelchli) [inline CHANGELOG.md:371]:
```suggestion
- Fixed an issue where class or init-only variables of dataclasses were passed to the dataclass constructor in `utilities.apply_to_collection` ([#9702](https://github.com/PyTorchLightning/pytorch-lightning/issues/9702))
```

**Comment by tchaton** [comment]
Dear @a-gardner1,

Would you mind resolving the conflicts ?

Best,
T.C

File: pytorch_lightning/utilities/apply_func.py (hunk 4 above)
Comment (a-gardner1) [reply pytorch_lightning/utilities/apply_func.py]: Cool. I didn't realize a `deepcopy` was acceptable. Not all tests run locally for me, but I think the latest version resolves the problem.

**Comment by carmocca** [comment]
Approved the CI to run - you can ping me on Slack if you need it again. It's no longer necessary after the first contribution.

**Comment by a-gardner1** [comment]
I cannot reproduce the failed test. To the best of my knowledge, it does not use anything that this PR has modified.

Local environment vs failed CI environment:
* CentOS 8.2.2004 vs Ubuntu 18.0.4, 
* Python 3.9.8 vs Python 3.9.7, 
* torch 1.10.0 vs torch 1.10.0

File: pytorch_lightning/utilities/apply_func.py [hunk 8]
```diff
@@ -119,21 +119,33 @@ def apply_to_collection(
...
-                    wrong_dtype=wrong_dtype,
-                    include_none=include_none,
-                    **kwargs,
-                )
-                if include_none or v is not None:
-                    out_dict[field.name] = v
-        return elem_type(**out_dict)
+            field_value = getattr(data, field.name)
+            fields[field.name] = (field_value, field.init)
+            memo[id(field_value)] = field_value
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:130]: nit: everything passes without the memo. do you think you could write a test to ensure it works properly?

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 9]
```diff
@@ -74,6 +74,11 @@ def _generate_sync_fn(self) -> None:
         # ignore typing, bad support for `partial`: mypy/issues/1484
         self._fn: Callable = partial(fn, reduce_op=self.op, group=self.group)  # type: ignore [arg-type]
 
+    def __deepcopy__(self, memo: dict) -> "_Sync":
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:77]: Can you add a comment about why this is necessary?

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 9 above)
Comment (a-gardner1) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:77]: In trying to find a satisfactory explanation, I discovered this change was just patching a symptom. The setter for `_Metadata.sync` alters the reduce operation for the `_Sync` object but does not regenerate its `_fn` attribute ([the problematic setter invocation](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/result.py#L429)). This (presumed) bug was masked in previous versions by [the call to `self.register_key`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/result.py#L433), which implicitly created a new `_Sync` object and regenerated `_fn` due to the way dataclasses were handled in `apply_to_collection`. Changing to a (deep)copy (forced by the inability to infer default values for `InitVar` constructor arguments) meant that this regeneration no longer occurred and that the `partial`ly applied `_fn` retained a `None` reduce op. 

I will be pushing a more targeted fix that eschews the custom `__deepcopy__` shortly.

File: pytorch_lightning/utilities/apply_func.py (hunk 8 above)
Comment (a-gardner1) [reply pytorch_lightning/utilities/apply_func.py:130]: The memo is meant to be an internal optimization that has no effect on the external API or observable functionality. However, there was one way in which this internal detail leaked: through non-init fields, which were deepcopied if the `memo` was omitted and passed through unchanged otherwise. The latest version rectifies this by ensuring that original non-init field values are explicitly set in the result.

With this change, I do not believe there is any reasonable test that could expose the usage of the memo.
